src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+# 
+# ```math
+# \left\{ \begin{aligned} 
+# \left[-\Lambda\nabla\cdot\frac{1}{\varepsilon(x)}\Lambda\nabla -k^2\right] H &= f(x) & \text{ in } \Omega,\\
+# H&=0 & \text{ on } \Gamma_D,\\
+# H|_{\Gamma_L}&=H|_{\Gamma_R},&\\
+# \end{aligned}\right.
+# ```
+# 
+# For convenience, in the weak form and Julia implementation below we represent $\Lambda$ as a vector instead of a diagonal $2 \times 2$ matrix, in which case $\Lambda\nabla$ becomes the elementwise product.
+# 
+# ## Topology Optimization
+# 
+# We use the density-based topology optimizaiton (TO) to optimize the electric field intensity at center. In TO, every point in the design domain is a design degree of freedom, which we discretize into a piece-wise constant parameter space $P$ for the design parameter $p\in [0,1]$. The material electric permittivity can then be determined by
+# 
+# ```math
+# \varepsilon(p) = \left[n_{air}+p(n_{metal}-n_{air})\right]^2, 
+# ```
+# where $n_{air}=1$ and $n_{metal}$ is the refractive index of the air and metal. Note that we do not directly interpolate the electric permittivity in the design region using the electric permittivity of air and metal, because the metal permittivity is negative and the direct interpolate might yield some divergence [4]. 
+# 
+# In practice, to avoid obtaining arbitrarily fine features as the spatial resolution is increased, one needs to impose a non-strict minimum lengthscale $r_f$. There are two schemes for this smoothing filter [5], here we perform the smoothing by solving a simple "damped diffusion" PDE, also called a Helmholtz filter [5]: 
+# ```math
+# \begin{aligned}    
+# -r_f^2\nabla^2p_f+p_f&=p\, ,\\
+# \left. \frac{\partial p_f}{\partial \vec{n}} \right\vert_{\partial\Omega_D} & =0 . 
+# \end{aligned} 
+# ```
+#
+# In this setup, we would have $r_f=R_f/(2\sqrt{3})$ where $R_f$ is the filter radius in another scheme used in the paper that we would want to compare with eventually [6]. 
+# 
+# Next, one employs a smooth threshold projection on the intermediate variable $p_f$ to obtain a "binarized" density parameter $p_t$ that tends towards values of $0$ or $1$ almost everywhere [6]: 
+# ```math
+# p_t = \frac{\tanh(\beta\eta)+\tanh\left[\beta(p_f-\eta)\right]}{\tanh(\beta\eta)+\tanh\left[\beta(1-\eta)\right]}. 
+# ```
+# Note that as $\beta\to\infty$, this threshold procedure goes to a step function, which would make the optimization problem non-differentiable. In consequence, the standard approach is to gradually increase $\beta$ to slowly binarize the design as the optimization progresses [6]. We will show how this is done below.  
+# 
+# Finally, we replace $p$ with the filtered and thresholded new parameter $p_t$ in the material permittivity expression above:
+# 
+# ```math
+# \varepsilon(p_t) = \left[n_{air}+p_t(n_{metal}-n_{air})\right]^2,
+# ```
```

Comment:
```suggestion
# ```
# This is the quantity that will be used for the $1/\varepsilon(x)$ coefficient in our Helmholtz PDE.
```